B(1) = +½ set 1: analytically correct Hurwitz zeta

Parcly-Taxel opened the pull request:
#### References to other Issues or PRs

This is part 1 of the changes formerly included in #23926.

#### Brief description of what is fixed or changed

mpmath's implementation of the Hurwitz zeta function agrees with Mathematica's `HurwitzZeta[]`, but the symbolic code in SymPy currently parallels `Zeta[]` instead. This PR rectifies the discrepancy, especially since `HurwitzZeta[]` is the correct analytic continuation of the Riemann zeta function and thus has the nicer properties.

#### Other comments

`bernoulli()` is also refactored in this PR, but its user-facing behaviour is not changed; `bernoulli(1)` still returns `-1/2`.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* functions
    * BREAKING: `zeta(s, a)` now implements the analytically continued Hurwitz zeta function (Mathematica's `HurwitzZeta[]`), not Mathematica's `Zeta[]`.
    * `zeta()` now simplifies for more sets of parameters.
<!-- END RELEASE NOTES -->

Parcly-Taxel committed:
functions: zeta() is HurwitzZeta[], not Zeta[]

In Mathematica there are two two-argument extensions of the Riemann zeta
function, Zeta[s,a] and HurwitzZeta[s,a]. The latter function provides
the correct analytic continuation of the Dirichlet sum to the left
half-plane in a, and thus satisfies a key identity in Apostol's
Introduction to Analytic Number Theory (Theorem 12.13 in the 1995
edition): when n is a non-negative integer, for *any* complex a

           -B     (a)
             n + 1
ζ(-n, a) = ───────────
              n + 1

where B denotes the Bernoulli polynomial, e.g. ζ(-1, -4) = -121/12.
There are other nice properties of HurwitzZeta[] which the other
extension Zeta[] lacks. Indeed, Zeta[] mangles the Dirichlet sum on the
left half-plane in a to the point where the graph looks more like
EllipticF[] than Log[].

SymPy's dependency mpmath agrees with HurwitzZeta[], but SymPy itself
has relations implementing Zeta[], so that zeta(-1, -4) = 119/12. This
commit rectifies the discrepancy.

With this change in behaviour also comes a refactoring of the evaluation
tree: ζ(s, a) can be analytically continued to all s ≠ 1 for any a not a
nonpositive integer, while Apostol's identity gives meaningful values to
the holes when s *is* a nonpositive integer. Thus ζ(s, a) is undefined
iff a is a nonpositive integer and s is not. The new evaluation tree
also suppresses explicit a = 1; to effect this change bernoulli()'s
evaluation tree had to be refactored in the same way, but its behaviour
is for the time being unchanged.

Parcly-Taxel committed:
Two imports

oscarbenjamin commented:
> * BREAKING: `zeta(s, a)` now implements the analytically continued Hurwitz zeta function (Mathematica's `HurwitzZeta[]`), not Mathematica's `Zeta[]`.

Does anyone else have any opinions on this? I think this change looks good to merge.

Parcly-Taxel commented:
> Does anyone else have any opinions on this? I think this change looks good to merge.

There is, however, a nagging question.

```
>>> zeta(4,5)
-22369/20736 + pi**4/90
```

Later on I am going to extend `harmonic()` to complex arguments using the zeta function:
```
harmonic(n, m) = zeta(m) - zeta(m, n+1)
```
This will be used for `harmonic.evalf()` and `harmonic.rewrite(zeta)`. However I currently have an automatic rewrite rule in `zeta.eval()` performing essentially the reverse transformation
```
zeta(s, a) -> zeta(s) - harmonic(a-1, s)
```
_if and only if `a` is an atomic integer greater than 1_, in which case the harmonic part collapses into a finite sum. Should I remove this rewrite of `zeta()` from automatic evaluation, leaving it to `zeta.expand_func()`?

oscarbenjamin commented:
> but frankly I think that recursive headaches will result if said behaviour remains.

These kinds of automatic rewrites are often problematic. Actually I now see that this PR extends this to non-integer `s` (unlike master):
```python
In [1]: zeta(S.Half, 2)
Out[1]: -harmonic(1, 1/2) + ζ(1/2)
```
I think it would be better at least not to extend this automatic rewrite.

Using this rewrite when both the zeta and the harmonic would fully evaluate so that we don't have any zeta or harmonic functions in the final result probably makes sense. Otherwise I don't think it's a good idea when the arguments are symbolic because of the recursive situations that you describe and also because automatic simplification should only apply when it is unambiguously a "good" simplification.

Parcly-Taxel committed:
functions: Restrict automatic zeta evaluation

In order to prevent potential infinite recursion when harmonic() is
extended to complex arguments, the scope of zeta(s,a)'s automatic
evaluation when a is an integer has to be restricted. Here the rewrite
is performed if and only if s is an Integer and a is a positive Integer,
ensuring that the harmonic part collapses into a Rational.

Parcly-Taxel commented:
> I think it would be better at least not to extend this automatic rewrite.

I should have sussed out that issue. I have decided that the automatic harmonic rewrite will be applied only if `s` and `a` are Integers.

oscarbenjamin commented:
Okay, I think this is good. Let's get it in. Thanks @Parcly-Taxel!

asmeurer commented:
Yes, the best pattern to follow is to only automatically evaluate on explicit Rational inputs, and leave more general symbolic evaluations to other simplification routines like `rewrite`, `doit`, etc. See https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval. We currently don't do a great job of following this in SymPy, but I'd like to move towards it (see also https://github.com/sympy/sympy/issues/23701).

> Does anyone else have any opinions on this? I think this change looks good to merge.

I'm curious if @jksuom has any thoughts on this and the underlying issue https://github.com/sympy/sympy/issues/23866

jksuom commented:
I would like to understand why the sign was changed in the first place but otherwise I see no problems with this.

asmeurer commented:
I think the sign was always -1/2 in SymPy. If you mean why it differed historically see the Luschny page linked in the issue.
